contrib/recipes/lease.go

```diff
@@ -0,0 +1,128 @@
+// Copyright 2016 CoreOS, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package recipe
+
+import (
+	"sync"
+	"time"
+
+	"github.com/coreos/etcd/Godeps/_workspace/src/golang.org/x/net/context"
+	"github.com/coreos/etcd/Godeps/_workspace/src/google.golang.org/grpc"
+	pb "github.com/coreos/etcd/etcdserver/etcdserverpb"
+	"github.com/coreos/etcd/lease"
+)
+
+// only keep one ephemeral lease per connection
+var connLeases connLeaseMgr = connLeaseMgr{leases: make(map[*grpc.ClientConn]*leaseKeepAlive)}
+
+type connLeaseMgr struct {
+	leases map[*grpc.ClientConn]*leaseKeepAlive
+	mu     sync.Mutex
+}
+
+type leaseKeepAlive struct {
+	id    lease.LeaseID
+	donec chan struct{}
+}
+
+func SessionLease(conn *grpc.ClientConn) (lease.LeaseID, error) {
+	return connLeases.sessionLease(conn, 120)
+}
+
+func SessionLeaseTTL(conn *grpc.ClientConn, ttl int64) (lease.LeaseID, error) {
+	return connLeases.sessionLease(conn, ttl)
+}
+
+// StopSessionLease ends the refresh for the session lease.
+func StopSessionLease(conn *grpc.ClientConn) {
+	connLeases.mu.Lock()
+	lka, ok := connLeases.leases[conn]
+	if ok {
+		delete(connLeases.leases, conn)
+	}
+	connLeases.mu.Unlock()
+	if lka != nil {
+		lka.donec <- struct{}{}
+		<-lka.donec
+	}
+}
+
+// RevokeSessionLease revokes the session lease.
+func RevokeSessionLease(conn *grpc.ClientConn) (err error) {
```

Comment:
I feel we should have the ability to revoke a specific lease by a given leaseID.